nixos/modules/services/misc/anki-sync-server.nix

```diff
@@ -0,0 +1,152 @@
+              description = lib.mdDoc ''
+                File containing the password accepted by anki-sync-server for
+                the associated username.  Make sure to make readable only by
+                root.
+              '';
+            };
+          };
+        });
+      description = lib.mdDoc "List of user-password pairs to provide to the sync server.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = builtins.length cfg.users > 0;
+        message = "At least one username-password pair must be set.";
+      }
+      {
+        assertion =
+          (
+            builtins.compareVersions
+            (elemAt (builtins.match ".*-(([0-9]+\\.)*[0-9]+)$" cfg.package.name) 0)
+            "2.1.66"
+          )
+          != -1;
+        message = "anki package <2.1.66 not supported by this module. Earliest supported commit: https://github.com/NixOS/nixpkgs/commit/05727304f8815825565c944d012f20a9a096838a";
+      }
+    ];
+    networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [cfg.port];
+
+    systemd.services.anki-sync-server = {
+      description = "anki-sync-server: Anki sync server built into Anki";
+      after = ["network.target"];
+      wantedBy = ["multi-user.target"];
+      path = [cfg.package];
+      environment =
+        {
+          SYNC_BASE = "%S/%N";
+          SYNC_HOST = specEscape cfg.host;
+          SYNC_PORT = toString cfg.port;
+        }
+        // (
+          attrsets.mergeAttrsList
+          (lists.imap1
+            (i: user: {"SYNC_USER${toString i}" = ''${specEscape user.username}:${
+                if user ? passwordFile
+                then "%d/${specEscape user.username}"
+                else specEscape user.password
+              }'';})
+            cfg.users)
+        );
+
+      serviceConfig = {
+        Type = "simple";
+        DynamicUser = true;
+        StateDirectory = name;
+        ExecStart = "${pkgs.bash}/bin/bash -c ${utils.escapeSystemdExecArg ''
+            # When services.anki-sync-server.users.passwordFile is set above,
+            # we set the "password" in the SYNC_USER* environment variables
+            # to the path of the password file, as exposed to the service.
+            # Since Anki sync server requires passwords to be passed via
+            # environment variable, here we replace each path with the content
+            # of the password file before launching the server.
+            user_vars=$(env | grep -oE '^SYNC_USER[0-9]+')
+            for var in ''$user_vars; do
+              eval "val=\''${''$var}";
+              user=''${val%%:*}
+              pass=''${val#*:}
+              # Only replace passwords that are paths (that begin with `/`).
+              if ! [[ "''$pass" = "''${pass##*/}" ]]; then
+                eval ''$var=''$(printf '%q' "''$user"):''$(printf '%q' ''$(cat "''$pass"))
+              fi
+            done
+            exec ${cfg.package}/bin/anki --syncserver
```

Comment:
Instead of embedding this directly in the unit, would it make more sense to use `writeScriptBin` for this and just use that with `ExecStart`? Makes it easier to run the same code outside of systemd.

Reply:
Done